File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        if(nums.length == 1){
            return nums[0];
        }
        else if(nums.length == 2){
            return Math.max(nums[0], nums[1]);
        }
        else if(nums.length == 3){
            return Math.max(nums[0]+nums[2], nums[1]);
        }
        nums[2] = nums[2] + nums[0];
        int max = nums[2];
        for(int i = 3; i<nums.length; i++){
            nums[i] = nums[i] + Math.max(nums[i-2], nums[i-3]);
            max = Math.max(max, nums[i]);
        }
        return max;
    }
}

File: 0198-house-robber/0198-house-robber.py
class Solution:
    def helper(self, nums: List[int], dp: List[int], index: int) -> int:
        if index == 0:
            dp[index] = nums[index]
            return dp[index]
        if index == 1:
            dp[index] = max(nums[0], nums[1])
            return dp[index]
        if dp[index] != -1: return dp[index]
        dp[index] = max(self.helper(nums, dp, index-1), nums[index] + self.helper(nums, dp, index-2))
        return dp[index]        
    def rob(self, nums: List[int]) -> int:
        dp = [-1 for i in range(len(nums))]
        return self.helper(nums, dp, len(nums)-1)